### ISSUE TITLE
Non-relevant section template data is sent with nonrelevant = "remove"

### ISSUE THREAD

**Issue by ebruchez** [orbeon/orbeon-forms#3829]
Properties:

```xml
<property as="xs:string"  name="oxf.fr.detail.buttons.acme.test-send-relevance">
    home summary review pdf send send
</property>

<property as="xs:string"  name="oxf.fr.detail.process.send.acme.test-send-relevance">
    send(
        uri         = "/fr/service/custom/orbeon/echo",
        replace     = "all",
        nonrelevant = "remove"
    )
</property>
```

[+1 from customer](https://3.basecamp.com/3600924/buckets/1966556/messages/1442328175)

**Comment by ebruchez** [comment]
The reason is that eventually, this makes it to `xf:submission/@nonrelevant` processing in `XFormsModelSubmissionBase`. And there, the way this is tested is with:

```
InstanceData.getInheritedRelevant
```

Since data mirrored from section templates does not mirror MIPs, the non-relevant nodes do not show as such. So we would need a solution for this.

**Comment by ebruchez** [comment]
Possible solutions:

- mirror MIPs, at least in → out
    - problem 1: MIPs are transient
    - problem 2: there is no MIP mutation event the `MirrorListener` could use right now
- listen to `xforms-enabled` / `xforms-disabled` in the section template implementation
    - add/remove `fr:relevant` (or similar) attributes on the inner instance
    - these would be mirrored out
    - `xf:submission` must handle these attributes as relevance data

**Comment by ebruchez** [comment]
One issue with the events approach is to know, when the component initializes, which controls are non-relevant. Currently, I don't believe that we dispatch events to non-relevant controls during the first refresh.

Also, an implementation note: we should ideally not duplicate all implementation code between section templates. Instead, shared code should be, well, shared.

**Comment by ebruchez** [comment]
We could also add an event for MIP changes (or at least for `relevant` MIP changes). Right now:

- `applyComputedExpressionBinds()` iterates binds recursively
- `handleComputedExpressionBind()` computes `relevant` (and `readonly` and `required`) as needed
- reminder
    - data nodes point to `InstanceData`, which itself points to `BindNode`s
    - local `relevant` is obtained by looking at all `BindNode`s
    - inherited `relevant` is obtained by checking all ancestor-or-self nodes

Now what we'd like to know is not whether an individual `relevant` MIP has changed, but if its effective ("inherited") value has changed. We'd need an algorithm to check that efficiently. In general, any node in the tree can change relevance during that process, in no particular order. Even nodes whose direct relevance hasn't changed could change inherited relevance.

(Also, in theory, once a data element is non-relevant, we shouldn't need to go check its descendants. But the hierarchy of binds doesn't necessarily follow the hierarchy of data (although it does for forms created by Form Builder). We could detect that the hierarchy is ordered, or order it when possible, but it's not the case right now.)

In short: there is no clear shortcut to obtain relevance change for a node. We could gather the relevance of all nodes before, and compare it with the relevance after, and dispatch events accordingly (maybe optionally, only for some instances). But is this a good enough algorithm?

**Comment by ebruchez** [comment]
Currently, we do not dispatch `xforms-disabled` upon tree creation. We only dispatch them:

- when the control was relevant and is no longer relevant
- when a repeat iteration is removed

Could we consider dispatching `xforms-disabled` to controls when the tree is created? Do we even like this, conceptually? If not, we could dispatch another event, such as `xxforms-initially-disabled`. This would prevent clashes with existing event listeners, and could also be easily optimized out when no listeners are present.

If we have `xxforms-initially-disabled`, then annotating with attributes becomes a real possibility.

**Comment by ebruchez** [comment]
Implementing `xxforms-initially-disabled` is fairly easy.

We'd like to reduce the possible impact on performance, though. If there is no handler, we optimize the event away, statically checking that there is no handler for the given event, which is good. Still, we are going to create event objects, but maybe we can optimize that a little bit as well.

The handlers would be within section templates, and we wouldn't need top-level handlers, probably - that is, if each section template gets its own handler. We could also imagine a model or instance attribute such as `xxf:dispatch-xxforms-initially-disabled` which would only dispatch such events if the control is in the scope of that model. Mabye.

**Comment by ebruchez** [comment]
For reference, right now:

- `fr-get-document-submission` clears `@xxf:relevant, @fr:relevant` upon load.
-  We annotate the data with `fr:relevant="false"` in `annotateNonRelevantElements` when preparing the data to save with `fr-create-update-submission`. If the node is relevant, we remove the attribute, otherwise we annotate and stop drilling down.

**Comment by ebruchez** [comment]
Issue: `event('xxf:binding')` returns an empty binding for a non-relevant control.

- Non-relevant controls get, in `ControlBindingSupport`, a `bindingContext`, no matter what (relevant/non-relevant, "new" binding or not).
- `ControlBindingSupport.binding`
    - returns `Nil` by default
    - is overridden by `XFormsSingleNodeControl` (`boundItem`) and `XFormsRepeatControl`
    - is used by
        - `XFormsComponentControl` for getting the `referenceNode` for mirroring
        - `event('xxf:binding')`
        - `xxf:binding()`
- `boundItem`
    - is obtained with `bc.getSingleItem` if `bc.newBind`
    - set in `XFormsSingleNodeControl` upon `onCreate()`/`onBindingUpdate()`
    - so if the control is non-relevant, it is not set

We could either modify `xxf:binding` to handle the non-relevant case as well:

- `XFormsComponentControl` doesn't need to use `binding`
- so we could modify `binding`?

**Comment by ebruchez** [comment]
For example, if the section enclosing the section template is initially relevant, then made non-relevant, and the section template has a non-relevant field, then the `fr:relevant` annotation remains in the data. At submission time, right now, the nested annotation would stay.

- [ ] `blankNonRelevantNodes`
    - [ ] would benefit from optimization
    - [ ] must prune extra annotations
- [x] `pruneNonRelevantNodes` is ok
- [ ] `annotateNonRelevantElements`
    - [ ] must remove nested annotations

**Comment by ebruchez** [comment]
There is indeed a bug if the section template is initially non-relevant. The tree of controls is still created, and that's where `xxforms-initially-disabled` is dispatched. However, the event handler is non-relevant, and that's why it doesn' run.

There is no `xforms-disabled` either, as the control was and is non-relevant.

Even if the event handler is relevant, the listener will not work if the section template is non-relevant. So we need a different way to handle this.

**Comment by ebruchez** [comment]
If the handlers are in the section template model, they are called, but there is no local instance to handle and mirror.

A possible solution: upon construction of the section template model, we annotate all elements which will get `xforms-enabled` when the controls refresh:

- [x] section containers
- [ ] repeated grid containers
    - but nothing is bound to them yet
- [x] section/grid iterations
- [ ] leaf controls

The trick is to properly identify those. We might have to follow the binds, as we do for simple data migration.

**Comment by ebruchez** [comment]
[Documented `xxf:relevant-attribute`](https://doc.orbeon.com/xforms/submission/submission-extensions#using-attributes-to-indicate-relevance-with-xxf-relevant-attribute)
